### ISSUE TITLE
Export setValue in useAsyncFn

### ISSUE THREAD

**Issue by Acidic9** [streamich/react-use#1462]
**Is your feature request related to a problem? Please describe.**
I want to modify the `state.value` from `useAsyncFn` (eg remove item from array) but I am unable to do so.

**Describe the solution you'd like**
`useAsyncFn` should support an additional item `setValue` which sets the value without needing to call the async function.

Example:
```js
const [state, callback, setState] = useAsyncFn(Promise.resolve([1, 2, 3]), []);

const removeItem = (value) => {
  const itemIndex = state.value.findIndex(item => item === value);
  const newState = state.slice().splice(itemIndex, 1);
  setState(newState);
}
```

**Describe alternatives you've considered**
Copy the source code of `useAsyncFn` and export a function to set the value.

**Comment by ngobach** [comment]
I don't think it is a good idea to have such feature, since the state of hook is not determinable.
For example, when the dependencies of the hook changed, then the state will be recalculated and any changes that was made via `setState` will be loss.

**Comment by Acidic9** [comment]
@ngobach I understand, but do you have a proposed alternative solution?

Would you recommend copying the value of `state.value` into a new `React.useState`? Or something else?

**Comment by ngobach** [comment]
Since the context is not clear. I cannot recommend you anything.
But it seems that using `useAsyncFn` in combination with `React.useState` could fit your expectation.

**Comment by benneq** [comment]
@ngobach He did provide a simple context:
1. async fetch a list of items, display the list, each item has a "delete" button
2. when pressing "delete": remove the item from the list

Of course you could re-fetch the whole list after deleting an item. But often that's an unnecessary API request.

Btw: I don't get the "not determinable" thing. I just created a copy of `useAsyncFn` and changed the last line to `return [state, (callback as unknown) as T, set];`. And then use it like that:
```
const [state, fetch, setState] = useAsyncFn(async () => {
  return await someAsyncFunction(dep1, dep2);
}, [dep1, dep2], { loading: false, value: [] });

const handleDelete = (itemIndex: number) => {
  setState(oldState => ({ ...oldState, value: oldState.value.filter((_, index) => index !== itemIndex) }));
};
```

I didn't encounter any unexpected behavior. It just works like `useState` with an additional `fetch` function.

**Comment by ngobach** [comment]
Hi @benneq,

I said I need more context because thread owner didn't mention the case when `useAsyncFn` re-executed, then it will overwrite any changes made by `setState` or they should merge together. It's like you have something with 2 sources of truth.

Let says ...

1. `useAsyncFn` gave me list of `[A, B, C]`.
2. setState changed the list to `[A, C]`.
3. `useAsyncFn` was re-run.
  a. The result should be `[A, B, C, D]` as output of the Promise from `useAsyncFn`, or
  b. The result should be `[A, C, D]` as output of the Promise from `useAsyncFn`, merging with B was being filtered out from step 2.

Maybe I was a bit over engineered.

**Comment by Acidic9** [comment]
In the case that the async fn was called again, it should override the state, so result b `[A, B, C]` should be the state. No merging should occur.

My point was simply that the function to set the state be exported as @benneq explained quite well - I don't really see exactly what issue this could cause. Also for performance, it might be good to avoid having to copy the state with another `React.useState`.

**Comment by benneq** [comment]
@ngobach That's an interesting thought.

I have one use case where merging the async data could be useful: infinite scroll pagination. Where each `fetch` will concat / merge / append the fetch results. But this makes things rather complicated.

In general I'd say: Don't change the existing behavior of the hook. Currently `useAsyncFn` **always** does override `state`. There's **no** merge.

Basically `setState` should behave like a "(synchronous) instant `fetch`".

---

Maybe I found another solution for this behavior (not tested):
```ts
const [state, fetch] = useAsyncFn(async (filter, sort, pagination, syncData) => {
  if(syncData) {
    return syncData;
  } else {
    return await asyncRequest(filter, sort, pagination);
  }
});
```

But that makes things much more complicated, because now there are 3+1 function arguments, which don't make sense together. (At least it should work as expected (?)). But still: A simple `setState` function would be much easier:
```ts
const [state, fetch, setState] = useAsyncFn(async (filter, sort, pagination) => {
  return await asyncRequest(filter, sort, pagination);
});

const foo = (syncData) => {
  setState(oldState => ({ loading: false, value: syncData }));
}
```
